syntax = "proto3";

// [#proto-status: draft]

package envoy.service.auth.v2alpha;
option go_package = "v2alpha";
option java_generic_services = true;

import "envoy/type/http_status.proto";
import "envoy/service/auth/v2alpha/attribute_context.proto";

import "google/rpc/status.proto";
import "validate/validate.proto";

// A generic interface for performing authorization check on incoming
// requests to a networked service.
service Authorization {
  // Performs authorization check based on the attributes associated with the
  // incoming request, and returns status `OK` or not `OK`.
  rpc Check(CheckRequest) returns (CheckResponse);
}

message CheckRequest {
  // The request attributes.
  AttributeContext attributes = 1;
}

// Http attributes for a denied response.
message DeniedHttpResponse {
  // Http status code.
  envoy.type.HttpStatus status = 1 [(validate.rules).message.required = true];

  // Http entity headers.
  map<string, string> headers = 2;
  
  // Http entity body.
  string body = 3;
}

// Http attributes for an ok response.
message OkHttpResponse {
  // Http entity headers in addition to the original request headers.
  map<string, string> headers = 2;
}

message CheckResponse {
  // Status `OK` allows the request. Any other status indicates the request should be denied.
  google.rpc.Status status = 1;

  // An message that contains HTTP response attributes. This message is
  // used when the authorization service needs to send custom responses to the
  // downstream client or, to modify/add request headers being dispatched to the upstream.
  oneof http_response {
    
    // Supplies http attributes for denied responses.
    DeniedHttpResponse denied_response = 2;
    
    // Supplies http attributes for ok responses.
    OkHttpResponse ok_response = 3;
  }
}
